### DISCUSSION TITLE
What is the proper way to select a Sandbox?

### DISCUSSION THREAD

**Discussion by shp7724** [ocamllabs/vscode-ocaml-platform#894] [Q&A] [answered]
Hi,

I'm completely new to `Ocaml` and trying to set up ocaml environment for vscode.
Especially I'm having a hard time with setting up this vscode extension to detect `ocamllsp`.

```bash
$ opam switch
#  switch   compiler      description
→  default  ocaml.4.12.0  default
```

```bash
$ which ocamllsp
/Users/shp7724/.opam/default/bin/ocamllsp
```

I assume this has to do with `Sandbox` and the issue #430. However, I couldn't find any documentation on how to set the opam sandbox to `default`.

Obviously the following setting does not work.

```
{
  "ocaml.sandbox": {
    "kind": "custom",
    "switch": "default"
  }
}
```

What is the proper way to select the `default` sandbox?
Thank you.

**Comment by mlantas** [comment]
The correct JSON would be
```json
"ocaml.sandbox": {
  "kind": "opam",
  "switch": "default"
}
```

However, the intended way to change the sandbox is through the command or status bar item: https://github.com/ocamllabs/vscode-ocaml-platform#setting-up-the-extension-for-your-project

**Reply by shp7724** [reply]
I tried both way (JSON and the command), but still it prints these warnings:

```
An error occurred starting the language server `ocamllsp`. Sandbox initialization failed: `ocaml-lsp-server` is not installed in the current sandbox.
```

```
OCaml bytecode compiler `ocamlc` was not found in the current sandbox. Do you have OCaml installed in the current sandbox?
```

I tried:

1. **Command Palette** > **Ocaml: Select a Sandbox for this Workspace** > **Global**: does not work.
2. **Command Palette** > **Ocaml: Select a Sandbox for this Workspace** > **Custom**: have no idea what to input as a custom command template.

Is there anything else that I can try? Or can you provide some relevant docs on *custom command template*?

**Reply by mlantas** [reply]
I think your environmental variables aren't being set correctly. You can try running `opam init` in a terminal which should prompt you to update your shell configuration.

<br/>

For a custom command template it shows you a default value:
![image](https://user-images.githubusercontent.com/25037249/156675569-149a47b5-64bb-41bc-abc3-9ea2f13ec8e6.png)

To make it use absolute paths you could try changing it to `/Users/shp7724/.opam/default/bin/$prog $args`. However, you will have a better integrated experience with an opam sandbox.

**Reply by shp7724** [reply]
I'm afraid I've already run `opam init` before.
Could you tell me how to verify my environmental variables setup?

These are the warnings that I have.

![스크린샷 2022-03-04 오후 11 12 37](https://user-images.githubusercontent.com/33917774/156778712-731a779e-aea5-4b00-b0dc-9800af20ced8.png)

**Reply by mlantas** [reply]
You're right, it looks like your `.zshrc` file already has the necessary command.

Can you please paste the contents of `OCaml Platform Extension` output channel here?
Do <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd> and use the `OCaml: Show OCaml Platform Extension Output` command to open the output channel.

**Comment by shp7724** [comment, marked as answer]
The problem was that I was using [code-server](https://github.com/coder/code-server), which is basically a fork of vscode, and for some reason the environment variables are ignored when the server is started.

So I added following lines to the startup script of `code-server` and everything worked as expected.

```bash
# OCaml Platform Extension Configuration
[[ ! -r /Users/shp7724/.opam/opam-init/init.zsh ]] || source /Users/shp7724/.opam/opam-init/init.zsh  > /dev/null 2> /dev/null
export PATH="/Users/shp7724/.opam/default/bin:/usr/local/bin:$PATH"
```

@mnxn Thanks for helping me find the solution.

**Reply by mlantas** [reply]
I'm not very familiar with `code-server`, so I'm not sure why that happens, but I'm glad you found a workaround.

Let us know if you have any other questions.

**Reply by Khady** [reply]
I'm guessing that the problem is the same as with vscode remote ssh. code-server is probably started in a non interactive session.

you can see in ~/.opam/opam-init/init.sh there are some lines protected by an if block.
```
if [ -t 0 ]; then
  test -r /home/louis.roche/.opam/opam-init/complete.sh && . /home/louis.roche/.opam/opam-init/complete.sh > /dev/null 2> /dev/null || true

  test -r /home/louis.roche/.opam/opam-init/env_hook.sh && . /home/louis.roche/.opam/opam-init/env_hook.sh > /dev/null 2> /dev/null || true
fi

test -r /home/louis.roche/.opam/opam-init/variables.sh && . /home/louis.roche/.opam/opam-init/variables.sh > /dev/null 2> /dev/null || true
```

By default during opam init, opam will modify `~/.profile` to source `~/.opam/opam-init/init.sh`. It works fine when creating an interactive session with  `ssh louis@myserver`. But it doesn't work flawlessly with vscode's remote ssh mode. What I mean by that is if I remote ssh with vscode on my server and open a shell inside vscode, then inside this shell navigate to a folder with a local switch, I get a warning from opam
```
[WARNING] The environment is not in sync with the current switch.
          You should run: eval $(opam env)
```
I don't have this warning when I ssh by hand on the server.

It can be solved by sourcing ~/.opam/opam-init/init.sh from ~/.bashrc instead of ~/.profile.
